### PR TITLE
feat(system-e2e): Move system-e2e to NX, v3

### PR DESCRIPTION
**Pull request by AndesKrrrrrrrrrrr**
#13968 

## Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Formatting passes locally with my changes
- [ ] I have rebased against main before asking for a review

## Summary by CodeRabbit

- **New Features**
  - Added a simple dummy test in `example.spec.ts` to verify Playwright functionality.
  - Implemented test configuration settings in `playwright.config.ts`.

- **Documentation**
  - Streamlined system testing instructions in `README.md`.
  - Improved test clarity and consistency across various test files.

- **Bug Fixes**
  - Corrected environment resolution logic in `urls.ts`.
  - Updated test scripts for proper Playwright tool handling.

- **Refactor**
  - Restructured custom matchers in `addons.ts` for better modularity.
  - Enhanced logging functionality and function syntax in `utils.ts`.

- **Chores**
  - Updated `.gitignore` to include Playwright test results directories.
  - Improved dependencies in `package.json` with latest tools and packages.

### CODE REVIEW

File: apps/system-e2e/e2e/example.spec.ts [hunk 1]
```diff
@@ -0,0 +1,8 @@
+import { test, expect } from '@playwright/test'
+
+test('has title', async ({ page }) => {
```
Comment (AndesKrrrrrrrrrrr) [inline apps/system-e2e/e2e/example.spec.ts:3]:
```suggestion
import { test, expect } from '@playwright/test'

// Very simple dummy test to verify that tests in `e2e/` also pass,
// also used for a _minimal_ tests to verify that Playwright works at all
test('has title', async ({ page }) => {
```

File: apps/system-e2e/entrypoint.sh [hunk 2]
```diff
@@ -14,24 +14,23 @@ if [[ "$*" =~ --project ]]; then
 fi
 
 export TEST_PROJECT TEST_ENVIRONMENT TEST_TYPE TEST_RESULTS_S3 TEST_FILTER
-
-DIR="$(cd "$(dirname "${BASH_SOURCE[0]}")" >/dev/null 2>&1 && pwd)"
+export PATH="./node_modules/.bin:$PATH"
```
Comment (AndesKrrrrrrrrrrr) [inline apps/system-e2e/entrypoint.sh:17]: Is this smart? Might lead to opaque issues :thinking: 

However, this removes the script's dependency on `yarn`

File: apps/system-e2e/entrypoint.sh [hunk 3]
```diff
@@ -14,24 +14,23 @@ if [[ "$*" =~ --project ]]; then
...
 
 echo "Current test environment: ${TEST_ENVIRONMENT}"
 echo "Playwright args: $*"
 echo "Playwright project: $TEST_PROJECT"
-echo "Playwright version: $(yarn playwright --version)"
+echo "Playwright version: $(playwright --version)"
 
 TEST_EXIT_CODE=0
-yarn playwright test -c src --project="$TEST_PROJECT" "$@" || TEST_EXIT_CODE=$?
+playwright test -c src --output test-results "$@" || TEST_EXIT_CODE=$?
```
Comment (AndesKrrrrrrrrrrr) [inline apps/system-e2e/entrypoint.sh:25]: Are you sure you want to remove `--project`? Do we really need the test output (via `--outptut`)?

File: apps/system-e2e/project.json
Comment (AndesKrrrrrrrrrrr) [inline apps/system-e2e/project.json:1]: Please remove unused targets.

File: apps/system-e2e/project.json [hunk 4]
```diff
@@ -9,34 +9,72 @@
     "build": {
       "executor": "nx:run-commands",
       "options": {
-        "commands": [
-          "esbuild --bundle $(find apps/system-e2e -name '*.ts' -not -path '*/node_modules/*') --outdir=dist/apps/system-e2e --tsconfig=apps/system-e2e/tsconfig.json --platform=node $(jq -r '.dependencies|keys[]|(\"--external:\"+.)' apps/system-e2e/package.json | xargs) --external:@nestjs/microservices --external:@nestjs/websockets --external:fsevents --external:class-transformer --external:canvas",
-          "cp apps/system-e2e/package.json dist/apps/system-e2e/"
-        ],
-        "outputPath": "dist/apps/system-e2e",
+        "commands": ["bash {projectRoot}/build.sh"],
```
Comment (AndesKrrrrrrrrrrr) [inline apps/system-e2e/project.json:12]: Do we _need_ `bash`? The build script uses bashism; can we change that?

File: apps/system-e2e/src/support/utils.ts [hunk 5]
```diff
@@ -25,6 +25,14 @@ export function deleteMockPdf() {
...
+export const debug = (msg?: string | object, ...args: unknown[]) => {
+  if (msg) {
+    if (typeof msg === 'object') {
+      debuglog('system-e2e')('', msg, ...args)
+      return
+    }
+    debuglog('system-e2e')(msg, ...args)
+    return
+  }
+  debuglog('system-e2e')('debug enabled')
```
Comment (AndesKrrrrrrrrrrr) [inline apps/system-e2e/src/support/utils.ts:37]: Consider using our central logging, or simply using Winston directly instead.

File: infra/scripts/build-docker-container.sh
Comment (AndesKrrrrrrrrrrr) [inline infra/scripts/build-docker-container.sh:1]: Please refrain from changing unrelated `infra` files

File: package.json [hunk 6]
```diff
@@ -333,6 +333,8 @@
     "@nestjs/schematics": "10.0.1",
     "@nestjs/testing": "10.0.5",
     "@nx/cypress": "16.9.0",
+    "@nx/devkit": "16.9",
+    "@nx/esbuild": "16.9",
```
Comment (AndesKrrrrrrrrrrr) [inline package.json:337]: Please pin patch versions

File: scripts/ci/90_docker-playwright.sh
Comment (AndesKrrrrrrrrrrr) [inline scripts/ci/90_docker-playwright.sh:1]: Although related to your PR, this file change is unnecessary.

File: scripts/ci/_docker.sh
Comment (AndesKrrrrrrrrrrr) [inline scripts/ci/_docker.sh:1]: Please refrain from changing unrelated `infra` files.

File: scripts/ci/ci.sh
Comment (AndesKrrrrrrrrrrr) [inline scripts/ci/ci.sh:1]: Please refrain from changing unrelated `infra` files.

File: apps/system-e2e/entrypoint.sh (hunk 3 above)
Comment (AndesKrrrrrrrrrrr) [reply apps/system-e2e/entrypoint.sh:25]: I can't get the outputs to go into `test-results` :shrug:

Comment (AndesKrrrrrrrrrrr) [reply apps/system-e2e/entrypoint.sh]: Removed `--output` usage. There are no more projects, so this should be safe. If we're still somewhere passing in `--project` for the container, the `entrypoint.sh` parses it out for us (and, now, disregards).

File: apps/system-e2e/build.sh [hunk 7]
```diff
@@ -0,0 +1,66 @@
+#!/bin/bash
+
+#
+# This script is used to build the system-e2e app using esbuild.
+# Playwright requieres special handling with esbuild, which is why it can't use
+# the "normal" build process.
+#
```
Comment (AndesKrrrrrrrrrrr) [inline apps/system-e2e/build.sh:7]: Even this seems to be failing when running against the built output. Playwright can't find our `expect` extensions like `expect.toHaveCountGreaterThan(...)`. It works locally, however, both when tested directly and with the `--ui` flag.

File: apps/system-e2e/entrypoint.sh [hunk 8]
```diff
@@ -14,24 +14,28 @@
...
+playwright test -c src "$@" || TEST_EXIT_CODE=$?
 
 # Upload results
 if [[ -n "$TEST_RESULTS_S3" ]]; then
   zip -r -0 test-results playwright-report src/test-results
   aws s3 cp test-results.zip "$TEST_RESULTS_S3"
 fi
 if [ "$TEST_EXIT_CODE" != "0" ]; then
-  yarn node "$DIR/src/notifications/notify.js"
+  node ./src/notifications/notify.*
```
[the remark this answers is not in the document]
Comment (AndesKrrrrrrrrrrr) [reply apps/system-e2e/entrypoint.sh:38]: If it fails, it fails. The only purpose of the script is to notify in our Slack channel. If something goes wrong, we won't get a notification, the same happens no matter how gracefully we handle all errors in the script.

**Review by lodmfjord — APPROVED** [review]
LGTM

File: apps/system-e2e/playwright.config.ts [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+            'tesults-build-description': process.env.COMMIT_INFO_MESSAGE,
+          },
+        ],
+      ]
+    : undefined,
+
+  projects: undefined, // Disable preset recommendation
+
+  /* Run your local dev server before starting the tests */
+  // webServer: {
```
Comment (saevarma) [inline apps/system-e2e/playwright.config.ts:48]: ~ Remove commented code?

File: apps/system-e2e/playwright.config.ts [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+
+// For CI, you may want to set BASE_URL to the deployed application.
+// const baseURL = process.env['BASE_URL'] || 'http://localhost:3000'
+const baseURL: string = urls.islandisBaseUrl
+
+/**
+ * Read environment variables from file.
+ * https://github.com/motdotla/dotenv
+ */
+// require('dotenv').config();
```
Comment (saevarma) [inline apps/system-e2e/playwright.config.ts:14]: Remove commented code?

File: apps/system-e2e/build.sh
Comment (saevarma) [inline apps/system-e2e/build.sh:1]: What special handling does Playwright requrie?
Did you look into if using javascript config to configure esbuild options dynamically would suffice?

I experimented with that when testing nx/esbuild with @busla for nx cloud. You can find reference here:
* https://github.com/island-is/island.is/blob/ci/nx-cloud/apps/services/sessions/esbuild.config.js
* https://github.com/island-is/island.is/blob/ci/nx-cloud/apps/services/sessions/project.json#L15
* https://github.com/island-is/island.is/blob/ci/nx-cloud/libs/shared/esbuild/esbuild.config.js

**Review by saevarma — COMMENTED** [review]
I wonder about the change here of "hardcoding" all labels in tests, that before were imported from the SUT messages file. This make e2e tests easily go out of sync. I'm interested in hearing your reason for removing this?

File: apps/system-e2e/build.sh
Comment (AndesKrrrrrrrrrrr) [reply apps/system-e2e/build.sh:1]: https://playwright.dev/docs/test-typescript#tsconfigjson

Playwright doesn't support includes or anything clever. They suggest building separately if we need anything more than basic.

I'm out of my depth here, and don't really care _how_ the tests end up running as long as they run and pass/fail correctly. Your expertise and guidance is highly appreciated :pray:

**Comment by AndesKrrrrrrrrrrr** [comment]
> I wonder about the change here of "hardcoding" all labels in tests

@saevarma, due to how playwright does importing, I was getting build errors, and onboarding was difficult. I plan to move all the tests to their respective app, where they can easily use relative-import for i18n.

However, for now, I need this (very unfortunately). I couldn't figure out the build, even though I spent a lot of time on it, and had help; I just think that path-aliased NX monorepo import isn't supported by Playwright.

File: apps/system-e2e/build.sh (hunk 7 above)
Comment (AndesKrrrrrrrrrrr) [reply apps/system-e2e/build.sh:7]: Removed all custom matchers :heavy_check_mark:

**Review by saevarma — APPROVED** [review]
🚀

**Pull request merged**